src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerJsonProcessor.java

```diff
@@ -239,6 +243,94 @@ public Content buildProviderJson(final Repository repository,
         ContentTypes.APPLICATION_JSON));
   }
 
+  /**
+   * Merges an incoming set of packages.json files.
+   */
+  public Content mergePackagesJson(final Repository repository, final List<Payload> payloads) throws IOException {
+    Set<String> names = new HashSet<>();
+    for (Payload payload : payloads) {
+      Map<String, Object> json = parseJson(payload);
+      Map<String, Object> providers = (Map<String, Object>) json.get(PROVIDERS_KEY);
+      names.addAll(providers.keySet());
+    }
+    return buildPackagesJson(repository, names);
+  }
+
+  /**
+   * Merges incoming provider JSON files, producing a merged file containing only the minimal subset of fields that we
+   * need to download artifacts.
+   */
+  public Content mergeProviderJson(final Repository repository, final List<Payload> payloads, final DateTime now)
+      throws IOException
+  {
+    String time = now.withZone(DateTimeZone.UTC).toString(timeFormatter);
+
+    // TODO: Make this more robust, right now it makes a lot of assumptions and doesn't deal with bad things well,
+    // can probably consolidate this with the handling for rewrites for proxy (or at least make it more rational).
+    Map<String, Map<String, Object>> packages = new LinkedHashMap<>();
+    for (Payload payload : payloads) {
+      Map<String, Object> json = parseJson(payload);
+      if (json.get(PACKAGES_KEY) instanceof Map) {
+
+        Map<String, Object> packagesMap = (Map<String, Object>) json.get(PACKAGES_KEY);
+        for (String packageName : packagesMap.keySet()) {
+
+          Map<String, Object> packageVersions = (Map<String, Object>) packagesMap.get(packageName);
+          for (String packageVersion : packageVersions.keySet()) {
+
+            Map<String, Object> versionInfo = (Map<String, Object>) packageVersions.get(packageVersion);
+            Map<String, Object> distInfo = (Map<String, Object>) versionInfo.get(DIST_KEY);
+            if (distInfo == null) {
+              continue;
+            }
+
+            Map<String, Object> newDistInfo = new LinkedHashMap<>();
+            newDistInfo.put(URL_KEY, String
+                .format(REWRITE_URL, repository.getUrl(), packageName, packageVersion, packageName.replace('/', '-'),
+                    packageVersion));
+            newDistInfo.put(TYPE_KEY, distInfo.get(TYPE_KEY));
+
+            Map<String, Object> newPackageInfo = new LinkedHashMap<>();
+            newPackageInfo.put(NAME_KEY, versionInfo.get(NAME_KEY));
+            newPackageInfo.put(VERSION_KEY, versionInfo.get(VERSION_KEY));
+            newPackageInfo.put(DIST_KEY, newDistInfo);
+            newPackageInfo.put(TIME_KEY, time);
+            newPackageInfo.put(UID_KEY, Integer.toUnsignedLong(
+                Hashing.md5().newHasher()
+                    .putString(packageName, StandardCharsets.UTF_8)
```

Comment:
Minor: As we do this twice a utility method that takes var args would make it clearer i.e. calculateMd5HashInteger(String... input);

Reply:
> Minor: As we do this twice a utility method that takes var args would make it clearer i.e. calculateMd5HashInteger(String... input);

Consolidated to a single usage as part of https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/14/commits/5ae09130b771886d44e76225f04a88dcd86965cf.